primo2/inference/dynamic.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# This file is part of PRIMO2 -- Probabilistic Inference Modules.
+# Copyright (C) 2013-2017 Social Cognitive Systems Group, 
+#                         Faculty of Technology, Bielefeld University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the Lesser GNU General Public License as 
+# published by the Free Software Foundation, either version 3 of the 
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public 
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+
+import abc
+import copy
+import six
+
+from . import exact
+
+
+__all__ = [
+    'PriorFeedbackExact',
+    'SoftEvidenceExact'
+]
+
+
+@six.add_metaclass(abc.ABCMeta)
+class DBNInferenceMethod(object):
+    """Abstract class for dynamic Bayesian network inference methods.
+
+    Contains the inference algorithm and the current state of the 
+    network at time t.
+    """
+
+    def __init__(self, dbn):
+        """Create the inference method and its state.
+
+        Parameters
+        ----------
+        dbn : DynamicBayesianNetwork
+            A 2-TBN dynamic Bayesian network structure.
+        """
+        super(DBNInferenceMethod, self).__init__()
+        self._dbn = dbn
+        self._t = 0
+        self._evidence = False
+
+    @abc.abstractmethod
+    def marginal_probabilities(self, variable):
```

Comment:
Maybe it would be more reasonable to let Factor be an internally used class only and change the existing inference methods to not return factors but only marginal probabilities or cpts in general. I mainly returned Factors since they were the objects coming out of the calculations and contained everything, but I guess it is confusing to ask for "potentials" when one is interested in probabilities.